### PR TITLE
PFM 1097 - Create / Update attendance register

### PR DESCRIPTION
**Pull request by chetankolte1**
1. Create attendance register
2. Update attendance register

### CODE REVIEW

File: backend/attendance-service/src/main/java/org/egov/web/models/AttendanceRegisterRequest.java [hunk 1]
```diff
@@ -21,7 +23,7 @@ public class AttendanceRegisterRequest {
     private RequestInfo requestInfo = null;
 
     @JsonProperty("attendanceRegister")
-    private AttendanceRegister attendanceRegister = null;
+    private List<AttendanceRegister> attendanceRegister = null;
```
Comment (rushang7-eGov) [inline backend/attendance-service/src/main/java/org/egov/web/models/AttendanceRegisterRequest.java:26]: Have you made the corresponding change in the API contract (.yaml file)?

Comment (chetankolte1) [reply backend/attendance-service/src/main/java/org/egov/web/models/AttendanceRegisterRequest.java:26]: Yes, corresponding change has been made in API contract

File: backend/attendance-service/src/main/java/org/egov/service/EnrichementService.java [hunk 2]
```diff
@@ -1,31 +1,86 @@
...
+
+    public void enrichUpdateAttendanceRegister(AttendanceRegisterRequest attendanceRegisterRequest) {
+        RequestInfo requestInfo = attendanceRegisterRequest.getRequestInfo();
+        List<AttendanceRegister> attendanceRegistersInUpdateReq = attendanceRegisterRequest.getAttendanceRegister();
+
+        for (int i = 0; i < attendanceRegistersInUpdateReq.size(); i++) {
+            AttendanceRegisterSearchCriteria searchCriteria = AttendanceRegisterSearchCriteria.builder()
+                    .id(attendanceRegistersInUpdateReq.get(i).getId().toString())
+                    .tenantId(attendanceRegistersInUpdateReq.get(i).getTenantId()).build();
+            AttendanceRegister attendanceRegisterFromDB = attendanceRepository.getAttendanceRegister(searchCriteria).get(0);
```
[the remark this answers is not in the document]
Comment (chetankolte1) [reply backend/attendance-service/src/main/java/org/egov/service/EnrichementService.java]: Done. Refactored the code to make single DB call

File: backend/attendance-service/src/main/java/org/egov/service/AttendanceService.java [hunk 3]
```diff
@@ -64,12 +68,35 @@ public List<AttendanceRegister> searchAttendanceRegister(RequestInfoWrapper requ
...
+        producer.push(attendanceServiceConfiguration.getUpdateAttendanceRegisterTopic(), attendanceRegisterRequest);
 
-        AttendanceRegister attendanceRegister = attendanceRegisterRequest.getAttendanceRegister();
-        ResponseInfo responseInfo = responseInfoFactory.createResponseInfoFromRequestInfo(attendanceRegisterRequest.getRequestInfo(), true);
-        AttendanceRegisterResponse attendanceRegisterResponse = AttendanceRegisterResponse.builder().responseInfo(responseInfo).attendanceRegister(Collections.singletonList(attendanceRegister)).build();
-        return attendanceRegisterResponse;
+        return attendanceRegisterRequest;
     }
+
+    private List<AttendanceRegister> getAttendanceRegisters(AttendanceRegisterRequest request) {
```
[the remark this answers is not in the document]
Comment (chetankolte1) [reply backend/attendance-service/src/main/java/org/egov/service/AttendanceService.java]: No /_search will call searchAttendanceRegister(), I am calling getAttendanceRegisters() implicitly for update. Should I rename this to getAttendanceRegistersToUpdate() ?

[the remark this answers is not in the document]
Comment (rushang7-eGov) [reply backend/attendance-service/src/main/java/org/egov/service/AttendanceService.java]: Ok. This is fine then.

File: backend/attendance-service/src/main/java/org/egov/service/AttendanceService.java [hunk 4]
```diff
@@ -64,12 +68,35 @@ public List<AttendanceRegister> searchAttendanceRegister(RequestInfoWrapper requ
      * @param attendanceRegisterRequest
      * @return
      */
-    public AttendanceRegisterResponse updateAttendanceRegister(AttendanceRegisterRequest attendanceRegisterRequest) {
-        //TODO Returning Dummy Response
+    public AttendanceRegisterRequest updateAttendanceRegister(AttendanceRegisterRequest attendanceRegisterRequest) {
+        attendanceServiceValidator.validateUpdateAttendanceRegisterRequest(attendanceRegisterRequest);
+        List<AttendanceRegister> attendanceRegistersFromDB = getAttendanceRegisters(attendanceRegisterRequest);
+        attendanceServiceValidator.validateAttendanceRegisterResponse(attendanceRegisterRequest, attendanceRegistersFromDB);
```
[the remark this answers is not in the document]
Comment (chetankolte1) [reply backend/attendance-service/src/main/java/org/egov/service/AttendanceService.java]: Renamed to validateUpdateAgainstDB()

File: backend/attendance-service/src/main/java/org/egov/service/EnrichementService.java [hunk 5]
```diff
@@ -1,31 +1,101 @@
...
+        for (int i = 0; i < attendanceRegisters.size(); i++) {
+            AuditDetails auditDetails = attendanceServiceUtil.getAuditDetails(requestInfo.getUserInfo().getUuid(), null,true);
+            attendanceRegisters.get(i).setAuditDetails(auditDetails);
+            attendanceRegisters.get(i).setId(UUID.randomUUID());
+
+            if (registerNumbers != null && !registerNumbers.isEmpty()) {
+                attendanceRegisters.get(i).setRegisterNumber(registerNumbers.get(i));
+            }
+
+            StaffPermission staffPermission = StaffPermission.builder().userId(requestInfo.getUserInfo().getUuid())
```
[the remark this answers is not in the document]
Comment (chetankolte1) [reply backend/attendance-service/src/main/java/org/egov/service/EnrichementService.java]: Move the code snippet to service file

**Review by rushang7-eGov — APPROVED** [review]

**Pull request merged**